### DISCUSSION TITLE
Global Search function for Documents

### DISCUSSION THREAD

**Discussion by betapro** [sumatrapdfreader/sumatrapdf#3611] [Q&A]
Is it possible for Sumatra to include a "**Global Document Search**" function (say, to the right of "**Find**"), so that it displays a list of (PDF) documents retrieved from the search criteria, which could then be individually opened at leisure? If yes, could a developer write something like this and integrate it into Sumatra? And what (tools, etc.) would be required for implementation?

**Comment by GitHubRulesOK** [comment]
You could use Acrobat it is able to do that usin C+#& for PDF using gazillions of coded  heuristics. (not one programmer)

The problem is both simple and complex.
The simple part is decoding a PDF can be done by Windows Search across all its folders, the complex part is PDF does not use text in a conventional manner.

here let me show you some PDF basics
This is how Acrobat may export a simple text sample
![image](https://github.com/sumatrapdfreader/sumatrapdf/assets/12754913/f2e174ab-2649-48c3-a536-462452189262)

So it starts the questions how is text found in a PDF ? and that depends on the sources thus authors thus potentially 8 billion different users using hundreds of techniques and or applications.

So Searching is one character by one character at a time then combine those into lines. and order those lines in a human fashion perhaps right to left or top to down or bottoms up. Commonly characters can get mixed up.

Acrobat did well to shuffle the letter order.
Here is the same file extracted by Sumatra PDF (that is NOT an error  it is the way the file is constructed, so Actually SumatraPDF is more "accurate".


![image](https://github.com/sumatrapdfreader/sumatrapdf/assets/12754913/d07a49c8-6519-4f8e-bc41-64fbc8cbd081)

```
How are you ?
Hell ,o World !
```
Thus the first task is "sorting" letters into a "Human" perception of placements, which is not a simple task for all PDFs on the planet.   

My favourite is Poppler PDFtoText -Layout as it attempts to produce sets of lines very similar to what is seen by eyes on a page.

Here is the Sample file (ignore any warnings that it is seen as unfinished that's not the reason it plays up just me not adding the index in the tail) [HelloWorld.pdf](https://github.com/sumatrapdfreader/sumatrapdf/files/12328054/HelloWorld.pdf)

Open it in Any PDF viewer AND MS Notepad

[HelloWorld.pdf](https://github.com/sumatrapdfreader/sumatrapdf/files/12328054/HelloWorld.pdf)

![image](https://github.com/sumatrapdfreader/sumatrapdf/assets/12754913/80af4b73-c999-4797-b665-644e8fd28394)

**Reply by betapro** [reply]
If Sumatra can  already **Find** text successfully in a PDF, what's to prevent it from doing a **Global Document Search** through a designated folder (with optional sub-folders) and finding all PDFs that meet the search criteria? Can it be done? (Y/N)

**Reply by GitHubRulesOK** [reply]
It does enable Global Search
It optionally installs the filter for Microsoft system to do the search 
EXACTLY same as Acrobat and others

The difference is Acrobat teams add extra ability to return the Microsoft search in an extra application sidebar window. any 3rd Party Searcher will use SumatraPDF to do its searches. However those apps may come with a price Tag.
A very cost effective one (apart from Acrobat) is MythicSoft Agent Ransack (the free version is also good) 
NOTE the free ones are usually slow run time whereas you need those that store SumatraPDF (or others) FASTER iFilter index.

![image](https://github.com/sumatrapdfreader/sumatrapdf/assets/12754913/4fb34c98-0d3d-4346-bab8-20b83d223071)

Change word and iFilter finds less in one second
--
![image](https://github.com/sumatrapdfreader/sumatrapdf/assets/12754913/c466bd04-8308-4be5-b7b1-ce5bef6a1246)

![image](https://github.com/sumatrapdfreader/sumatrapdf/assets/12754913/ba1cffb8-2f09-4b2b-95ed-882be2f03d0d)

I aborted Acrobat after 1 minute as too slow so result for those 2 files is 
![image](https://github.com/sumatrapdfreader/sumatrapdf/assets/12754913/fa04506e-ebd8-479a-ab3a-73f60db496f7)


**The fastest way to do a Global Search** is click on a **saved one** and **INSTANTLY** see the results  

![image](https://github.com/sumatrapdfreader/sumatrapdf/assets/12754913/7e775637-ca29-496f-82fa-09d08ed40c6a)

Somewhere Microsoft have notes on how to call those from CMD or VBA or VBS or JS or WSH or other apps via their API. However for "Ad Hoc" fastest is to click a saved one OR for fresh type in the search boxes.

**Personally I have no problems on many faster devices** (more usage ?) using iFilters (**but do not use on slow devices**) for more about potential issues see https://github.com/sumatrapdfreader/sumatrapdf/discussions/2460 and https://github.com/sumatrapdfreader/sumatrapdf/issues/3372

Move a file or rename a copy or move to another folder WINDOWS will track the SEARCH changes
---
![image](https://github.com/sumatrapdfreader/sumatrapdf/assets/12754913/6290e305-f626-44e0-8117-c301229f7e56)
